working/2364 - primary constructors/feature-specification.md

```diff
@@ -361,14 +397,28 @@ constants *(holding parameter default values)*.
 Where no processing is mentioned below, _D2_ is identical to _D_. Changes
 occur as follows:
 
+Assume that `p` is an optional formal parameter in _D_ that does not have a
+declared type, but it does have a default value whose static type in the
+empty context is a type `T` which is not `Null`. In that case `p` is
+considered to have the declared type `T`. When `T` is `Null`, `p` is
+considered to have the declared type `Object?`.
+
 The current scope of the formal parameter list of the primary constructor
 in _D_ is the current scope of the class/enum declaration *(in other words,
 the default values cannot see declarations in the class body)*. Every
 default value in the primary constructor of _D_ is replaced by a fresh
 private name `_n`, and a constant variable named `_n` is added to the
 top-level of the current library, with an initializing expression which is
-said default value. *(This means that we can move the parameter
-declarations including the default value without changing its meaning.)*
+said default value. 
```

Comment:
Rewrites! Can we avoid them? 

I assume this rewrite happens because we also desugar the primary constructor into a synthetic syntactic declaration of a normal constructor, and need to preserve the scope of the moved expressions. That means moving expressions between scopes, and I can see why that's bad.
(If the expressions were in the parameter scope, inside the type parameter scope and the class body scope, like they normally would be for a constructor parameter default value, then moving them into that position wouldn't change anything. Maybe that's another reason to do that.)

Also, with assert expressions, the scope of those cannot be hoisted in the same way. They need access to the parameter scope (class name, type parameters, parameters), even if not the class body scope.

A rewrite is a change, which means there is a before and after.
*When* are the default value *expressions* replaced by the fresh identifier?
What happens before and what happens after.
(I'd guess it happens *very early*, right after parsing, because it can.)